### PR TITLE
feat(NODE-1400): Add zst compatibility for NNS subnet recovery

### PR DESCRIPTION
**Pull request by andrewbattat**
NODE-1400
We are switching to using zst images for upgrades so that we can retire gz ic-os images.

Upgrades are currently [tested on zst images](https://github.com/dfinity/ic/blob/master/rs/tests/driver/src/driver/test_env_api.rs)

The final piece of logic that must change is to make download_registry_and_restart_if_nns_subnet_recovery agnostic to tar type

### CODE REVIEW

File: rs/http_utils/src/file_downloader.rs [hunk 1]
```diff
@@ -182,16 +183,49 @@ pub fn check_file_hash(path: &Path, expected_sha256_hex: &str) -> FileDownloadRe
...
 }
 
-/// Extract the contents of a given .tar.gz file into `target_dir`
-pub fn extract_tar_gz_into_dir(tar_gz_path: &Path, target_dir: &Path) -> FileDownloadResult<()> {
-    let map_to_untar_error = |e| FileDownloadError::untar_error(tar_gz_path, e);
+/// Check if the file is in Gzip format by verifying the first 2 bytes
+fn is_gz_file<R: Read>(reader: &mut R) -> io::Result<bool> {
+    let mut buffer = [0; 2];
+    reader.read_exact(&mut buffer)?;
+    reader.by_ref().take(2).read_to_end(&mut Vec::new())?;
```
Comment (eichhorl) [inline rs/http_utils/src/file_downloader.rs:190]: Why is this line needed here (and in `is_zst_file`)?

Comment (andrewbattat) [reply rs/http_utils/src/file_downloader.rs]: You are correct—that line is not needed. Removed! Thank you

File: rs/http_utils/src/file_downloader.rs [hunk 2]
```diff
@@ -182,16 +183,47 @@ pub fn check_file_hash(path: &Path, expected_sha256_hex: &str) -> FileDownloadRe
...
+        File::open(tar_path).map_err(|e| FileDownloadError::file_open_error(tar_path, e))?;
+    let mut buf_reader = io::BufReader::new(tar_file);
+
+    if is_gz_file(&mut buf_reader).unwrap_or(false) {
+        let tar_gz_file =
+            File::open(tar_path).map_err(|e| FileDownloadError::file_open_error(tar_path, e))?;
+        let tar = GzDecoder::new(tar_gz_file);
+        let mut archive = Archive::new(tar);
+        archive.unpack(target_dir).map_err(map_to_untar_error)?;
+    } else if is_zst_file(&mut buf_reader).unwrap_or(false) {
```
Comment (kpop-dfinity) [inline rs/http_utils/src/file_downloader.rs:214]: I'm not sure how `BufReader` works exactly, but doesn't the stream position change when we call `is_gz_file` meaning that when we call `is_zst_file` after calling `is_gz_file` we will not read the very first 4 bytes of the file but instead the bytes at positions `2,3,4,5` in the file?

Comment (derlerd-dfinity) [reply rs/http_utils/src/file_downloader.rs:214]: +1 

I guess it might also be good to add a test along with the change to make sure the new functionality works as expected.

Comment (andrewbattat) [reply rs/http_utils/src/file_downloader.rs]: BufReader bug fixed and tests added! Thank you!

File: rs/http_utils/src/file_downloader.rs [hunk 3]
```diff
@@ -182,17 +184,58 @@ pub fn check_file_hash(path: &Path, expected_sha256_hex: &str) -> FileDownloadRe
...
+pub fn extract_tar_into_dir(tar_path: &Path, target_dir: &Path) -> FileDownloadResult<()> {
+    let map_to_untar_error = |e| FileDownloadError::untar_error(tar_path, e);
+
+    let tar_file =
+        File::open(tar_path).map_err(|e| FileDownloadError::file_open_error(tar_path, e))?;
+    let mut buf_reader = BufReader::new(tar_file);
+
+    // Check for gzip format
+    buf_reader
+        .seek(SeekFrom::Start(0))
```
Comment (frankdavid) [inline rs/http_utils/src/file_downloader.rs:211]: Is it really necessary to seek to the beginning of a file that you just opened for reading?

File: rs/http_utils/src/file_downloader.rs [hunk 4]
```diff
@@ -182,17 +184,58 @@ pub fn check_file_hash(path: &Path, expected_sha256_hex: &str) -> FileDownloadRe
...
+        .map_err(map_to_untar_error)?;
+    if is_gz_file(&mut buf_reader).map_err(map_to_untar_error)? {
```
Comment (frankdavid) [inline rs/http_utils/src/file_downloader.rs:213]: Have you considered moving the code for seeking to the beginning into the is_XXX functions? Then you could remove all the seeking from this function which would make it slightly easier to read. In addition, it's also less surprising (to me at least) if the check doesn't advance the cursor.

File: rs/http_utils/src/file_downloader.rs [hunk 5]
```diff
@@ -472,4 +520,83 @@ mod tests {
...
+        encoder.finish().unwrap();
+
+        extract_tar_into_dir(&tar_path, &extract_dir).unwrap();
+
+        let extracted_file = extract_dir.join("test.txt");
+        let mut file = File::open(extracted_file).unwrap();
+        let mut contents = String::new();
+        file.read_to_string(&mut contents).unwrap();
+        assert_eq!(contents, "Hello, world!");
+    }
```
Comment (frankdavid) [inline rs/http_utils/src/file_downloader.rs:601]: Consider adding a test case for the error branch.

File: rs/http_utils/src/file_downloader.rs [hunk 6]
```diff
@@ -472,4 +520,83 @@ mod tests {
...
+        let mut encoder = GzEncoder::new(tar_gz, Compression::default());
+        create_tar(&mut encoder).unwrap();
+        encoder.finish().unwrap();
+
+        extract_tar_into_dir(&tar_path, &extract_dir).unwrap();
+
+        let extracted_file = extract_dir.join("test.txt");
+        let mut file = File::open(extracted_file).unwrap();
+        let mut contents = String::new();
+        file.read_to_string(&mut contents).unwrap();
```
Comment (frankdavid) [inline rs/http_utils/src/file_downloader.rs:579]: nit: std::fs::read_to_string(extracted_file) does everything from lines 577-579 and is simpler (same below).

Comment (andrewbattat) [reply rs/http_utils/src/file_downloader.rs]: Fixed! Thank you!

File: rs/http_utils/src/file_downloader.rs (hunk 5 above)
Comment (andrewbattat) [reply rs/http_utils/src/file_downloader.rs:585]: Added! Thank you!

File: rs/http_utils/src/file_downloader.rs (hunk 4 above)
Comment (andrewbattat) [reply rs/http_utils/src/file_downloader.rs:213]: Great note! Seeks moved inside the is_XXX functions, much more readable now.

File: rs/http_utils/src/file_downloader.rs (hunk 3 above)
Comment (andrewbattat) [reply rs/http_utils/src/file_downloader.rs]: For the current code, it's not necessary, but now that the seeks are moved inside the is_XXX functions, I think it's better to have them there for robustness sake, in case they're ever called with the reader in an unexpected position.

**Comment by eichhorl** [comment]
> We are switching to using zst images for upgrades so that we can retire gz ic-os images.

As I understand, this functionality is currently only used in the NNS recovery with failover nodes case, where we create a `tar.gz` of the registry local store, which is then downloaded by the recovery failover nodes. Is there a specific reason that this should be switched to `zst`, and do you also plan to adapt the code generating these archives (see [rs/recovery](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/recovery/src/lib.rs?L782-797))?

**Review by nmattia — APPROVED** [review]

**Comment by andrewbattat** [comment]
> > We are switching to using zst images for upgrades so that we can retire gz ic-os images.
> 
> As I understand, this functionality is currently only used in the NNS recovery with failover nodes case, where we create a `tar.gz` of the registry local store, which is then downloaded by the recovery failover nodes. Is there a specific reason that this should be switched to `zst`, and do you also plan to adapt the code generating these archives (see [rs/recovery](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/recovery/src/lib.rs?L782-797))?

Ah, I see. I naively assumed that the recovery url from the CUP comes from a URL in a release package. Is this not true?

I tried looking in rs/recovery to see how exactly the registry_store_uri value is populated but I couldn't track it down. 

This change may not be necessary. Do you know how the registry_store_uri value is populated? Will the switch to using tar.zst images in release proposals not effect this code then? And is there any CI testing for the nns subnet recovery?

**Comment by eichhorl** [comment]
> I naively assumed that the recovery url from the CUP comes from a URL in a release package. Is this not true?

The `registry_store_uri` field holds the URL to a `tar.gz` (or `tar.zst` after your change) of the local registry store that should be deployed on the recovered NNS

> I tried looking in rs/recovery to see how exactly the registry_store_uri value is populated but I couldn't track it down.

It is set when sending the recovery CUP proposal [here](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/recovery/src/admin_helper.rs?L222). The file itself is generated by the function [here](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/recovery/src/lib.rs?L782).

> Will the switch to using tar.zst images in release proposals not effect this code then?

Release images are downloaded using the [download_file](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/orchestrator/image_upgrader/src/image_upgrader.rs?L165-167) function. We then call [manageboot.sh upgrade-install](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/orchestrator/image_upgrader/src/image_upgrader.rs?L200-209), which [unpacks](https://sourcegraph.com/github.com/dfinity/ic/-/blob/ic-os/components/upgrade/manageboot/manageboot.sh?L270) the image

> is there any CI testing for the nns subnet recovery?

There is the `sr_nns_failover_nodes_test` which creates and uploads the archive [here](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/tests/consensus/orchestrator/sr_nns_failover_nodes_test.rs?L265-277)

**Comment by andrewbattat** [comment]
@eichhorl Thank you for this info!

I have updated rs/recovery and sr_nns_same_nodes_test_colocate is passing! I'll merge this tomorrow, unless you say otherwise.

**Review by eichhorl — APPROVED** [review]
